File: src/test/java/com/example/hexagonal/adapter/in/ProductControllerTest.java
package com.example.hexagonal.adapter.in;

import static com.example.hexagonal.domain.ModelFixture.DEFAULT_ID;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

import com.example.hexagonal.port.dto.DtoFixture;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

@SpringBootTest
class ProductControllerTest {

  static ObjectMapper mapper = new ObjectMapper(); // Don't do in production; Declare a bean

  MockMvc mvc;

  @Test
  void getProduct() throws Exception {
    mvc
        .perform(get("/api/products/{productId}", DEFAULT_ID)
            .accept("application/json")
        )
        .andDo(print())
        .andExpect(status().is2xxSuccessful());
  }

  @Test
  void updateProduct() throws Exception {
    final byte[] content = mapper.writeValueAsBytes(DtoFixture.aProductDtoOf(DEFAULT_ID));

    mvc
        .perform(put("/api/products/{productId}", DEFAULT_ID)
            .contentType("application/json")
            .accept("application/json")
            .content(content)
        )
        .andDo(print())
        .andExpect(status().is2xxSuccessful());
  }

  @BeforeEach
  void setup() {
    this.mvc = MockMvcBuilders
        .standaloneSetup(new ProductController())
        .build();
  }
}

File: kotlin/src/test/kotlin/com/example/hexagonal/adapter/driving/ProductControllerTest.kt
package com.example.hexagonal.adapter.driving

import com.example.hexagonal.TestConfig
import com.example.hexagonal.domain.Product
import com.example.hexagonal.port.driving.ProductPort
import com.example.hexagonal.usecase.dto.DtoMapper
import com.fasterxml.jackson.databind.ObjectMapper
import io.kotest.core.spec.style.DescribeSpec
import io.kotest.extensions.spring.SpringExtension
import io.mockk.every
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.context.annotation.Import
import org.springframework.test.context.TestConstructor
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.ResultActions
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultHandlers
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.setup.MockMvcBuilders

@WebMvcTest(ProductController::class)
@Import(TestConfig::class, DtoMapper::class)
@TestConstructor(autowireMode = TestConstructor.AutowireMode.ALL)
internal open class ProductControllerTest(
    val mockUsecase: ProductPort,
    val dtoMapper: DtoMapper,
) : DescribeSpec() {
    override fun extensions() = listOf(SpringExtension)

    init {
        describe("When an API client calls 'getProduct' API") {
            it("should success with 2xx status") {
                getProduct()
            }
        }

        describe("When an API client calls 'updateProduct' API") {
            it("should success with 2xx status") {
                updateProduct()
            }
        }
    }

    open fun getProduct() {
        every { mockUsecase.getProduct(any()) } returns Product.aProduct()

        val res: ResultActions = mvc
            .perform {
                MockMvcRequestBuilders
                    .get("/api/products/{productId}", fixture.id)
                    .accept("application/json")
                    .buildRequest(it)
            }
            .andDo(MockMvcResultHandlers.print())

        res
            .andExpect(MockMvcResultMatchers.status().is2xxSuccessful())
    }

    open fun updateProduct() {
        val dto = Product.aProduct()
        every { mockUsecase.updateProduct(any(), any()) } returns dto

        val res: ResultActions = mvc
            .perform {
                MockMvcRequestBuilders
                    .put("/api/products/{productId}", fixture.id)
                    .contentType("application/json")
                    .accept("application/json")
                    .content(mapper.writeValueAsBytes(dto))
                    .buildRequest(it)
            }
            .andDo(MockMvcResultHandlers.print())

        res
            .andExpect(MockMvcResultMatchers.status().is2xxSuccessful())
    }

    val mvc: MockMvc = MockMvcBuilders
        .standaloneSetup(ProductController(mockUsecase, dtoMapper))
        .build()

    val fixture: Product = Product.aProduct()

    companion object {
        val mapper = ObjectMapper()
    }
}
